### DISCUSSION TITLE
How to disable Datasource on certain @MicronautTest-annotated test?

### DISCUSSION THREAD

**Discussion by Nahuel92** [micronaut-projects/micronaut-data#2856] [Q&A] [answered]
### Issue description

Hi there.

I'm using Micronaut Data in my application, but I would like to disable the Datasource for certain tests I have annotated with @MicronautTest. I haven't seen anything like that in the docs, where the suggested approach is to use a DB such as H2.

In my particular case, what I'm testing right now doesn't use the DB at all, but it still requires it if I want to use the Micronaut context to use its real dependencies (which, in turn, are not DB-related).

I would like to achieve something similar to test slices that can be achieved with Spring Boot so that I can use a context but without a DB (for certain tests).

Is that even possible?
Thanks.

**Comment by sdelamo** [comment]
I am moving this issue to Micronaut Data. 

One option is to remove `datasources.default.*` configuration from the default `application.properties` that will be loaded in tests. That will cause Micronaut Data not to be started.

**Reply by sdelamo** [reply]
In Micronaut SQL JPA we added https://micronaut-projects.github.io/micronaut-sql/5.5.1/guide/#hibernate-disable

**Reply by Nahuel92** [reply]
Hi Sergio, thanks for your prompt answer and for converting this issue into a discussion. Apologies for the delay, I have been having internet issues for the last couple of days. 

I forgot to mention that I'm using Micronaut Data JDBC. Maybe we could have a `jdbc.enabled` flag for that (or a more general `datasources.enabled`).
Regarding your suggestion, I already tried that using an `application-test.yml` without any `datasource.default.*` property, but Micronaut keeps picking the `application.yml` file and complains about the DB being down. For that specific test, I don't require one (I'm testing how a mapper and a service work together).

As, eventually, I will include a couple of tests that will exercise the whole application (it's small enough at this point to allow this for the main paths at least), I was looking for something to exclude DB initialization from tests like Spring Boot does with their test slices.

I may be taking the wrong path, and if that is true, I'll appreciate any help. Thanks again Sergio!!

**Reply by sdelamo** [reply, marked as answer]
Yes, move everything datasource related to outside of `application.properties`. 

In the test which use persistence annotate them with 

```java
@Property(name = "datasources.default.password", value = "")
@Property(name = "datasources.default.dialect", value = "H2")
@Property(name = "datasources.default.schema-generate", value = "CREATE_DROP")
@Property(name = "datasources.default.url", value = "jdbc\:h2\:mem\:devDb;LOCK_TIMEOUT\=10000;DB_CLOSE_ON_EXIT\=FALSE")
@Property(name = "datasources.default.username", value = "sa")
@Property(name = "datasources.default.driver-class-name", value = "org.h2.Driver")
@MicronautTest
...
..
```

and have: 

`application-dev.properties` to run the app locally with datasource related configuration an DEV as the [default environment](https://docs.micronaut.io/4.3.12/guide/#_default_environment).
`application-prod.properties` to configure production database. run your app in production with `MICRONAUT_ENVIRONMENTS=prod`

But, I agree we need `datasource.enabled` or `jdbc.enabled`.

**Reply by Nahuel92** [reply]
Thank you Sergio, that actually was an approach I tried but I'm still stuck in an old Micronaut version (3.7.x) and I couldn't make it work.

Unfortunately, when running locally from a dev machine, Micronaut won't pick up any environment, and it will read properties from the global application.yml instead of the application-dev.yml. The app refuses to start since there are a few required properties are not found.

Since I'm not the only dev, I'm trying to allow a good dev experience for the team. What I came up for now is to use Helm chart to inject Micronaut datasources.* properties (even though this is coupling the Helm chart with application implementation details) and commenting out those properties in the main application.yml.

That way, tests will work as long as a developer doesn't uncomment those properties to work with the application. It's a flaky and error-prone approach, but I guess until I can update to Micronaut 4.3.x this is the best I can come up with.

Any suggestion is more than welcomed. Thanks again Sergio!!

**Reply by eyalgo** [reply]
Hi,
We're facing the same issue with Micronaut 4.3.3 and I'm wondering if I miss anything, or should we solve it with _application-X.yml_ files as @sdelamo propsed?

Thanks

**Reply by sdelamo** [reply]
@eyalgo Currently, I recommend you to solve it with multiple application-env.properties files

**Reply by Nahuel92** [reply]
Sorry for the delay. Sergio's workaround works like a charm, and I decided to use it to solve my problem.
Thanks again Sergio!!

**Comment by sdelamo** [comment]
I created an issue to enable programmatic disabling https://github.com/micronaut-projects/micronaut-data/issues/2893
